File: src/test/java/no/nav/syfo/consumer/ws/ArbeidsfordelingConsumerTest.java
package no.nav.syfo.consumer.ws;

import no.nav.syfo.domain.dto.Soknadstype;
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.ArbeidsfordelingV1;
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.FinnBehandlendeEnhetListeUgyldigInput;
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.*;
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.meldinger.WSFinnBehandlendeEnhetListeRequest;
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.meldinger.WSFinnBehandlendeEnhetListeResponse;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static java.util.Collections.singletonList;
import static no.nav.syfo.consumer.ws.ArbeidsfordelingConsumer.BEHANDLINGSTEMA_OPPHOLD_UTLAND;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;


@RunWith(MockitoJUnitRunner.class)
public class ArbeidsfordelingConsumerTest {

    @Mock
    private ArbeidsfordelingV1 arbeidsfordelingV1;

    @InjectMocks
    private ArbeidsfordelingConsumer arbeidsfordelingConsumer;

    @Before
    public void setup() throws FinnBehandlendeEnhetListeUgyldigInput {
        when(arbeidsfordelingV1.finnBehandlendeEnhetListe(any()))
                .thenReturn(new WSFinnBehandlendeEnhetListeResponse().withBehandlendeEnhetListe(singletonList(
                        new WSOrganisasjonsenhet()
                                .withStatus(WSEnhetsstatus.AKTIV)
                                .withEnhetId("enhetsId"))));
    }

    @Test
    public void hentBehandlendeEnhet() throws FinnBehandlendeEnhetListeUgyldigInput {
        GeografiskTilknytning geografiskTilknytning = GeografiskTilknytning.builder()
                .geografiskTilknytning("2017")
                .diskresjonskode(null)
                .build();

        arbeidsfordelingConsumer.finnBehandlendeEnhet(geografiskTilknytning, Soknadstype.SELVSTENDIGE_OG_FRILANSERE);

        ArgumentCaptor<WSFinnBehandlendeEnhetListeRequest> captor = ArgumentCaptor.forClass(WSFinnBehandlendeEnhetListeRequest.class);

        verify(arbeidsfordelingV1).finnBehandlendeEnhetListe(captor.capture());

        WSArbeidsfordelingKriterier kriterier = captor.getValue().getArbeidsfordelingKriterier();

        assertThat(kriterier.getDiskresjonskode()).isNull();
        assertThat(kriterier.getBehandlingstema()).isNull();
        assertThat(kriterier.getGeografiskTilknytning()).isEqualTo(new WSGeografi().withValue("2017"));
        assertThat(kriterier.getTema()).isEqualTo(new WSTema().withValue("SYK"));
    }

    @Test
    public void hentBehandlendeEnhetKode6GirDiskresjonskode() throws FinnBehandlendeEnhetListeUgyldigInput {
        GeografiskTilknytning geografiskTilknytning = GeografiskTilknytning.builder()
                .geografiskTilknytning("2017")
                .diskresjonskode("SPSF")
                .build();

        arbeidsfordelingConsumer.finnBehandlendeEnhet(geografiskTilknytning, Soknadstype.OPPHOLD_UTLAND);

        ArgumentCaptor<WSFinnBehandlendeEnhetListeRequest> captor = ArgumentCaptor.forClass(WSFinnBehandlendeEnhetListeRequest.class);

        verify(arbeidsfordelingV1).finnBehandlendeEnhetListe(captor.capture());

        WSArbeidsfordelingKriterier kriterier = captor.getValue().getArbeidsfordelingKriterier();

        assertThat(kriterier.getDiskresjonskode()).isEqualTo(new WSDiskresjonskoder().withValue("SPSF"));
        assertThat(kriterier.getBehandlingstema()).isEqualTo(new WSBehandlingstema().withValue(BEHANDLINGSTEMA_OPPHOLD_UTLAND));
        assertThat(kriterier.getGeografiskTilknytning()).isEqualTo(new WSGeografi().withValue("2017"));
        assertThat(kriterier.getTema()).isEqualTo(new WSTema().withValue("SYK"));
    }
}

File: src/test/kotlin/no/nav/syfo/consumer/ws/ArbeidsfordelingConsumerTest.kt
package no.nav.syfo.consumer.ws

import no.nav.syfo.domain.dto.Soknadstype
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.binding.ArbeidsfordelingV1
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.binding.FinnBehandlendeEnhetListeUgyldigInput
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Behandlingstema
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Diskresjonskoder
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Enhetsstatus
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Geografi
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Organisasjonsenhet
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Tema
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.meldinger.FinnBehandlendeEnhetListeRequest
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.meldinger.FinnBehandlendeEnhetListeResponse
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers.any
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner


@RunWith(MockitoJUnitRunner::class)
class ArbeidsfordelingConsumerTest {

    @Mock
    private val arbeidsfordelingV1: ArbeidsfordelingV1? = null

    @InjectMocks
    private val arbeidsfordelingConsumer: ArbeidsfordelingConsumer? = null

    @Before
    @Throws(FinnBehandlendeEnhetListeUgyldigInput::class)
    fun setup() {
        `when`(arbeidsfordelingV1!!.finnBehandlendeEnhetListe(any()))
            .thenReturn(FinnBehandlendeEnhetListeResponse().apply {
                behandlendeEnhetListe.add(Organisasjonsenhet().apply {
                    status = Enhetsstatus.AKTIV
                    enhetId = "enhetsId"
                })
            })
    }

    @Test
    @Throws(FinnBehandlendeEnhetListeUgyldigInput::class)
    fun hentBehandlendeEnhet() {
        val geografiskTilknytning = GeografiskTilknytning("2017", null)

        arbeidsfordelingConsumer!!.finnBehandlendeEnhet(geografiskTilknytning, Soknadstype.SELVSTENDIGE_OG_FRILANSERE)

        val captor = ArgumentCaptor.forClass(FinnBehandlendeEnhetListeRequest::class.java)

        verify<ArbeidsfordelingV1>(arbeidsfordelingV1).finnBehandlendeEnhetListe(captor.capture())

        val kriterier = captor.value.arbeidsfordelingKriterier

        assertThat(kriterier.diskresjonskode).isNull()
        assertThat(kriterier.behandlingstema).isNull()
        assertThat(kriterier.geografiskTilknytning.value).isEqualTo("2017")
        assertThat(kriterier.tema.value).isEqualTo("SYK")
    }

    @Test
    @Throws(FinnBehandlendeEnhetListeUgyldigInput::class)
    fun hentBehandlendeEnhetKode6GirDiskresjonskode() {
        val geografiskTilknytning = GeografiskTilknytning("2017", "SPSF")

        arbeidsfordelingConsumer!!.finnBehandlendeEnhet(geografiskTilknytning, Soknadstype.OPPHOLD_UTLAND)

        val captor = ArgumentCaptor.forClass(FinnBehandlendeEnhetListeRequest::class.java)

        verify<ArbeidsfordelingV1>(arbeidsfordelingV1).finnBehandlendeEnhetListe(captor.capture())

        val kriterier = captor.value.arbeidsfordelingKriterier

        assertThat(kriterier.diskresjonskode.value).isEqualTo("SPSF")
        assertThat(kriterier.behandlingstema.value).isEqualTo(BEHANDLINGSTEMA_OPPHOLD_UTLAND)
        assertThat(kriterier.geografiskTilknytning.value).isEqualTo("2017")
        assertThat(kriterier.tema.value).isEqualTo("SYK")
    }
}
